components/needs_team/needs_team.jsx

```diff
@@ -81,35 +84,104 @@ export default class NeedsTeam extends React.Component {
+                this.joinTeam(nextProps);
+            }
         }
     }
 
-    updateCurrentTeam(props) {
-        // First check to make sure you're in the current team
-        // for the current url.
-        const teamName = props.match.params.team;
-        const team = TeamStore.getByName(teamName);
+    componentDidMount() {
+        startPeriodicStatusUpdates();
+        startPeriodicSync();
 
-        if (!team) {
-            props.history.push('/?redirect_to=' + encodeURIComponent(props.location.pathname));
-            return null;
+        // Set up tracking for whether the window is active
+        window.isActive = true;
+        Utils.applyTheme(this.props.theme);
+
+        if (UserAgent.isIosSafari()) {
+            // Use iNoBounce to prevent scrolling past the boundaries of the page
+            iNoBounce.enable();
         }
 
-        // If current team is set, then this is not first load
-        // The first load action pulls team unreads
-        if (TeamStore.getCurrentId()) {
-            this.props.actions.getMyTeamUnreads();
+        window.addEventListener('focus', this.handleFocus);
+        window.addEventListener('blur', this.handleBlur);
+        window.addEventListener('keydown', this.shortcutKeyDown);
+    }
+
+    componentDidUpdate(prevProps) {
+        const {theme} = this.props;
+        if (!Utils.areObjectsEqual(prevProps.theme, theme)) {
+            Utils.applyTheme(theme);
+        }
+    }
+
+    componentWillUnmount() {
+        stopPeriodicStatusUpdates();
+        stopPeriodicSync();
+        if (UserAgent.isIosSafari()) {
+            iNoBounce.disable();
         }
 
-        TeamStore.saveMyTeam(team);
-        BrowserStore.setGlobalItem('team', team.id);
-        TeamStore.emitChange();
+        clearInterval(wakeUpInterval);
+        window.removeEventListener('focus', this.handleFocus);
+        window.removeEventListener('blur', this.handleBlur);
+        window.removeEventListener('keydown', this.shortcutKeyDown);
+    }
+
+    handleBlur = () => {
+        window.isActive = false;
+        this.blurTime = new Date().getTime();
+        if (this.props.currentUser) {
+            this.props.actions.viewChannel('');
+        }
+    }
+
+    handleFocus = async () => {
+        await this.props.actions.markChannelAsRead(this.props.currentChannelId);
+        window.isActive = true;
+
+        if (new Date().getTime() - this.blurTime > UNREAD_CHECK_TIME_MILLISECONDS) {
+            this.props.actions.fetchMyChannelsAndMembers(this.props.currentTeamId).then(loadProfilesForSidebar);
+        }
+    }
+
+    joinTeam = async (props) => {
+        const openTeams = await this.props.actions.getTeams(0, TEAMS_PER_PAGE);
+        const team = openTeams.data.find((teamObj) => teamObj.name === props.match.params.team);
+        if (team) {
+            const {error} = await props.actions.joinTeam(team.invite_id, team.id);
+            if (error) {
+                props.history.push('/error_page');
```

Comment:
Instead of `/error_page`, should this be pushed to `/error` and/or pass params like type, etc?

Reply:
Waiting for feedback on this. Yes we should have params and define the states for this